### DISCUSSION TITLE
Command Line Syntax

### DISCUSSION THREAD

**Discussion by WinMerge2021** [WinMerge/winmerge#1675] [General]
I usually compare folders with command line syntax.  When I run like 10 of these in a row, they all open in the same WinMerge instance.

**Questions:**

Is it possible to open only those that have differences?

Or

if it does open, is it possible to close all tabs that are identical (or have NO differences)?

Or

Get a return code if it is identical or differences?

**In other words, when I run the command line syntax, I only care about the one that are different and then it needs to open in WinMerge.  If it is identical, just ignore and return a return code that it is identical.  Right now, I have to press Esc to close it.**

Thanks

**Comment by sdottaka** [comment]
It is only possible to return the comparison result with the process exit status code.

Using the command line options /enableexitcode and /noninteractive, WinMerge will exit as soon as the comparison is finished and the ERRORLEVEL environment variable will be set to the comparison result.
If the ERRORLEVEL is non-zero, then there is a difference and you should try to open WinMerge again.

~~~
start /w "" c:\progra~1\winmerge\winmergeu.exe /enableexitcode /noninteractive /r c:\dev\winmerge\src\ d:\dev\winmerge\src\
echo %ERRORLEVEL%
if not %ERRORLEVEL% == 0 c:\progra~1\winmerge\winmergeu.exe /r c:\dev\winmerge\src\ d:\dev\winmerge\src\
~~~

**Reply by WinMerge2021** [reply]
Thanks for your reply SDOTTaka.  Strange, I ran your first two lines of your code at commandline.

**`start /w "" "C:\Test\WinMergeU.exe" /enableexitcode /noninteractive /r "A:\" "B:\" | Echo %ERRORLEVEL%`**

Although I added one new file onto the A drive, the above returned a 0.  Exactly what I am looking for if it works.

**Reply by WinMerge2021** [reply]
ErrorLevel returned a 0 or a 1, if there is a difference.  So that part seems to work.  But ErrorLevel did not return the comparison result as you have indicated, such as the actual files or folders.  Is this correct?

**Reply by sdottaka** [reply]
If you remove ` | Echo %ERRORLEVEL%` and put `Echo %ERRORLEVEL%` on the next line, it will work as expected.
Or, if you want to run `Echo %ERRORLEVEL%` only if there is a difference, change it to ` || Echo %ERRORLEVEL%` instead of ` | Echo %ERRORLEVEL%`.

Note that if the value of ERRORLEVEL is 0, it is identical; if it is 1, there is a difference; if it is 2, it is an error.

**Comment by WinMerge2021** [comment]
Thanks SDotTaka.  Really appreciate your help as always.

This worked for me but I had to remove the space before Echo.  The `||` did not work.

`start /w "" "C:\Test\WinMergeU.exe" /enableexitcode /noninteractive /r "A:\" "B:\" |Echo %ErrorLevel%
`


Yes, I agree:

> Note that if the value of ERRORLEVEL is 0, it is identical; if it is 1, there is a difference; if it is 2, it is an error.

I might have misunderstood your explanation earlier and thought that it would return the comparison results in the form of the actual differences and items.  But I believe you mean the return code above right?

**Reply by sdottaka** [reply]
I thought I had tried `||` , but it didn't seem to work.

> I might have misunderstood your explanation earlier and thought that it would return the comparison results in the form of the actual differences and items. But I believe you mean the return code above right?

Yes, %ERRORLEVEL% indicates only difference or identical, or errors.

If you want detailed information on the differences, not just whether or not there are differences, create a report file. Below is an example of outputting to a CSV file.

https://github.com/WinMerge/winmerge/discussions/1341

**Reply by WinMerge2021** [reply]
You are pretty Awesome SdotTaka.  I have all the information that you gave me to automate my process.

Is there an update of all the Command Line Switches?
https://manual.winmerge.org/en/Command_line.html

What is `ReportFiles/ReportType` because it is not in the link above.

`/cfg ReportFiles/ReportType=0`

**Reply by sdottaka** [reply]
The content of the link above is up to date.

ReportFiles/ReportType represents the report type.
0: CSV, 1: TSV, 2: HTML, 3: XML

I'm hesitant to document it because I might use another way of specifying it in the future.
